File: 2024/04/go/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

type position struct {
	row int
	col int
}

var data = [][]rune{}
var posMap = map[rune]map[position]struct{}{}

func parse(r io.Reader) {
	scanner := bufio.NewScanner(r)

	for scanner.Scan() {
		data = append(data, []rune(scanner.Text()))
	}
}

func scan(rSlice ...rune) {
	runeMap := map[rune]struct{}{}
	for _, r := range rSlice {
		runeMap[r] = struct{}{}
	}

	for r := range runeMap {
		posMap[r] = map[position]struct{}{}
	}

	for i, row := range data {
		for j, cell := range row {
			if _, ok := runeMap[cell]; ok {
				posMap[cell][position{i, j}] = struct{}{}
			}
		}
	}
}

func linearSearch(w string) (res int) {
	vectors := [8][2]int{{1, 1}, {1, 0}, {1, -1}, {0, 1}, {0, -1}, {-1, 1}, {-1, 0}, {-1, -1}}
	for pos := range posMap[rune(w[0])] {
	vectorLoop:
		for _, v := range vectors {
			for i, r := range w {
				row := pos.row + v[0]*i
				if row >= len(data) || row < 0 {
					continue vectorLoop
				}

				col := pos.col + v[1]*i
				if col >= len(data[0]) || col < 0 {
					continue vectorLoop
				}

				if data[row][col] != r {
					continue vectorLoop
				}
			}

			res++
		}
	}

	return
}

func CrossSearch(center rune, corner1 rune, corner2 rune) (count int) {
	for pos := range posMap[center] {
		if pos.row == 0 || pos.col == 0 || pos.row == len(data)-1 || pos.col == len(data[0])-1 {
			continue
		}

		corners := []rune{}
		corners = append(corners, data[pos.row-1][pos.col-1])
		corners = append(corners, data[pos.row-1][pos.col+1])
		corners = append(corners, data[pos.row+1][pos.col+1])
		corners = append(corners, data[pos.row+1][pos.col-1])


		if (corners[0] == corners[1] && corners[2] == corners[3]) ||
			(corners[0] == corners[3] && corners[1] == corners[2]) {
			cornerMap := map[rune]struct{}{}
			for _, c := range corners {
				cornerMap[c] = struct{}{}
			}

			_, ok1 := cornerMap[corner1]
			_, ok2 := cornerMap[corner2]

			if ok1 && ok2 {
				count++
			}
		}
	}
	return
}
func main() {
	parse(os.Stdin)
	scan('X', 'A')

	P1 := linearSearch("XMAS")
	P2 := CrossSearch('A', 'M', 'S')

	fmt.Printf("P1: %d\n", P1)
	fmt.Printf("P2: %d\n", P2)
}

File: 2024/04/typescript/main.ts
import { readFileSync } from "fs";

const vectors = [[0, 1], [1, 0], [0, -1], [-1, 0]]
const diagonal_vectors = [[1, 1], [1, -1], [-1, -1], [-1, +1]]

class WordSearch {
    data: Array<Array<string>>

    constructor(d: Array<Array<string>>) {
        this.data = d
    }

    linearSearch(word: string, [row, col]: [number, number]): number {
        let res = 0
        l_vector: for (const [c_row, c_col] of vectors.concat(diagonal_vectors)) {
            for (let i = 0; i < word.length; i++) {
                const c = word[i]
                const [n_row, n_col] = [row + i * c_row, col + i * c_col]
                if ((n_row >= this.data.length || n_row < 0) || (n_col >= this.data[n_row].length || n_col < 0)) {
                    continue l_vector
                }

                if (this.data[n_row][n_col] === c) {
                    continue
                }

                continue l_vector
            }

            res++
        }

        return res
    }

    crossSearch([c1, c2]: [string, string], [row, col]: [number, number]): boolean {
        const corners = diagonal_vectors.map(([c_row, c_col]) =>
            this.data[row + c_row][col + c_col]
        )

        if (
            (
                (corners[0] === corners[1] && corners[2] === corners[3]) ||
                (corners[1] === corners[2] && corners[3] === corners[0])
            ) &&
            (
                (corners[0] === c1 && corners[2] === c2) ||
                (corners[0] === c2 && corners[2] === c1)
            )
        ) {
            return true
        }

        return false

    }
}

const input_data: Array<Array<string>> = []
for (const line of readFileSync(0, { encoding: "utf-8" }).split("\n")) {
    input_data.push([...line])
}

input_data.pop()

const data = new WordSearch(input_data)

let P1: number = 0
for (let row = 0; row < data.data.length; row++) {
    for (let col = 0; col < data.data[row].length; col++) {
        if (data.data[row][col] === "X") {
            P1 += data.linearSearch("XMAS", [row, col])
        }
    }
}

console.log(`P1: ${P1}`)

let P2: number = 0
for (let row = 1; row < data.data.length - 1; row++) {
    for (let col = 1; col < data.data[row].length - 1; col++) {

        if (data.data[row][col] === "A") {
            if (data.crossSearch(["M", "S"], [row, col])) {
                P2++
            }
        }
    }
}

console.log(`P2: ${P2}`)